x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java

```diff
@@ -137,6 +153,61 @@ public void testDatetime() throws IOException, SQLException {
         }
     }
 
+    public void testDatetimeWithNanos() throws IOException, SQLException {
+        assumeTrue("Driver version [" + JDBC_DRIVER_VERSION + "] doesn't support DATETIME with nanosecond resolution]",
+                versionSupportsDateNanos());
+
+        long randomTimestampWitnNanos = randomNanos();
+        int randomNanosOnly = extractNanosOnly(randomTimestampWitnNanos);
+        setupIndexForDateTimeTests(randomTimestampWitnNanos, true);
+
+        try (Connection connection = esJdbc()) {
+            try (PreparedStatement statement = connection.prepareStatement(
+                    "SELECT id, test_date_nanos FROM emps WHERE test_date_nanos = ?")) {
+                Timestamp ts = new Timestamp(toMilliSeconds(randomTimestampWitnNanos));
+                statement.setObject(1, ts);
+                try (ResultSet results = statement.executeQuery()) {
+                    assertFalse(results.next());
+                }
+
+                ts.setNanos(randomNanosOnly);
+                statement.setObject(1, ts);
+                try (ResultSet results = statement.executeQuery()) {
+                    assertTrue(results.next());
+                    assertEquals(1002, results.getInt(1));
+                    assertEquals(ts, results.getTimestamp(2));
+                    assertFalse(results.next());
+                }
+            }
+        }
+    }
+
+    public void testDateTimeWithNanosAgainstDriverWithoutSupport() throws IOException, SQLException {
+        assumeFalse("Driver version [" + JDBC_DRIVER_VERSION + "] doesn't support DATETIME with nanosecond resolution]",
+                versionSupportsDateNanos());
+
+        long randomTimestampWitnNanos = randomNanos();
+        int randomNanosOnly = extractNanosOnly(randomTimestampWitnNanos);
+        setupIndexForDateTimeTests(randomTimestampWitnNanos, true);
+
+        try (Connection connection = esJdbc()) {
+            try (PreparedStatement statement = connection.prepareStatement(
+                    "SELECT id, test_date_nanos FROM emps WHERE test_date_nanos = ?")) {
+                Timestamp ts = new Timestamp(toMilliSeconds(randomTimestampWitnNanos));
+                statement.setObject(1, ts);
+                try (ResultSet results = statement.executeQuery()) {
+                    assertFalse(results.next());
```

Comment:
We set nanos resolution to the `Timestamp` which are ignored by the older JDBC drivers, so since we filter on a `date_nanos` column (field) the millis only resolution that is used by the older JDBC  won't match the document.

Reply:
And the second assertion fails because the nanos resolution of the parameter is ignored by the driver and, in fact, the query will become `... WHERE test_date_nanos = [value_with_millis_resolution]`?

Reply:
exactly!